File: dto/src/main/kotlin/cz/vutbr/fit/knot/enticing/dto/response/FullDocument.kt
package cz.vutbr.fit.knot.enticing.dto.response

import cz.vutbr.fit.knot.enticing.dto.utils.Extended
import cz.vutbr.fit.knot.enticing.dto.utils.Extension
import cz.vutbr.fit.knot.enticing.dto.utils.regex.urlRegexStr
import javax.validation.Valid
import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.Pattern
import javax.validation.constraints.Positive

/**
 * Full document content returned from an IndexServer
 */
data class FullDocument(
        @field:NotEmpty
        val title: String,
        @field:NotEmpty
        @field:Pattern(regexp = urlRegexStr)
        val url: String,
        @field:Valid
        val payload: Payload.FullResponse,
        @field:Valid
        val queryMapping: List<QueryMapping> = emptyList()
)

/**
 * Information handled by Webserver
 */
data class DocumentExtra(
        /**
         * Url of the index server
         */
        @field:NotBlank
        @field:Pattern(regexp = urlRegexStr)
        val host: String,
        /**
         * Collection on the server
         */
        @field:NotEmpty
        val collection: String,
        /**
         * Id of the document in the collection
         */
        @field:Positive
        val documentId: Int,

        /**
         * Query, optional
         *
         * if provided, QueryMapping informing about the mapping between the query and the document will
         * be included in the response
         */
        val query: String? = null
) : Extension

typealias ExtendedDocument = Extended<FullDocument, DocumentExtra>

File: frontend/src/entities/FullDocument.ts
import {QueryMapping, queryMappingSchema} from "./Annotation";
import {Payload, payloadSchema} from "./Snippet";
import * as yup from "yup";
import {validateOrNull} from "./validationUtils";

export interface FullDocument {
    host: string
    collection: string
    documentId: number
    title: string
    url: string
    payload: Payload
    query?: string
    queryMapping: Array<QueryMapping>
}

const documentSchema = yup.object({
    host: yup.string().required(),
    collection: yup.string().required(),
    documentId: yup.number().integer().required(),
    title: yup.string().required(),
    url: yup.string().url().required(),
    payload: payloadSchema,
    query: yup.string(),
    queryMapping: yup.array(queryMappingSchema)
})

export function isDocument(obj: Object): obj is FullDocument {
    return validateOrNull(documentSchema, obj) !== null;
}


